devnet-sdk/testing/systest/multi_client.go

```diff
@@ -186,15 +186,54 @@ func (mc *MultiClient) fetchWithConsistencyCheck(
 		return nil, err
 	}
 
+	// Get the state root for the primary
+	primaryHeader, ok := primaryItem.(*types.Header)
+	if !ok {
+		// If we're fetching a block instead of a header
+		if block, ok := primaryItem.(*types.Block); ok {
+			primaryHeader = block.Header()
+		}
+	}
+
 	// Create a hash-only getter for followers
 	getFollowerHash := func(client HeaderProvider, num *big.Int) (common.Hash, error) {
-		_, _, hash, err := queryFn(client, num)
-		return hash, err
+		item, _, hash, err := queryFn(client, num)
+		if err != nil {
+			return common.Hash{}, err
+		}
+
+		// If hashes don't match, also check state roots
```

Comment:
I am not sure if we get anything actionable by knowing that state root matches despite divergent hashes.

As an engineer running this test, I anyway would be motivated to dig into this, the moment I observe unequal hashes irrespective of the state root.

That plus the rarity of this happening, I would be inclined to just having this test based on simply comparing the hashes to be fine and credible.

Reply:
Having debugged a lot of client divergences, it would be useful to print the full block header in the error message. It's very common for the state root to differ and that requires digging a lot deeper to find what differed in the state but there are many cases where things like the base fee adjustment or gas used differs and that shows up directly in the block header and can be a useful starting point.  It can be any field in the header and is often multiple fields, so I wouldn't try to assert multiple fields, but including the full header is very handy.  It could be json which is readable but verbose or RLP which is not immediately readable but more compact and can be easily parsed.

Reply:
@yashvardhan-kukreja - I've added this extra check to print a more tailored error message as per previous feedback from @geoknee .

@ajsutton , in the event the headers differ I now return an error that includes both. I haven't converted to JSON or RLP for now; will leave that for a future improvement so as to not increase the scope of this PR.

Reply:
@scharissis I think in the case where the state roots match but the block hashes diverge you are still only logging out the hashes. 

Honestly I think this code could be really simple: if the block hashes do not match then throw an error with the full block headers (if they are available). Then, if a divergence occurs whoever is debugging it has all the information they need. I don't think it is necessary to have the code start singling out fields (like the state root) to see if they match. This can be done by the user.